train.py

```diff
@@ -392,6 +413,42 @@ def loss_fn(pred, labels):
                 time_last_log = time.perf_counter()
                 gpu_memory_monitor.reset_peak_stats()
 
+            # log val metrics
+            if job_config.validation.enable_val and (
+                train_state.step == 0
+                or train_state.step % eval_freq == 0
+            ):
+                val_data_loader = build_hf_data_loader(
```

Comment:
Simplifies the interface at neglible performance cost, since we do all data processing on-the-fly. Otherwise, we will have to add a method to reset the dataloader state. Feel free to add this if you're willing to.